File: crates/flighthq-scene-wgpu/src/normal_wgpu_mesh_material_renderer.rs
//! The built-in Normal forward mesh-material renderer — the WGSL mirror of
//! `normal_gl_mesh_material_renderer`.
//!
//! Ports `@flighthq/scene-wgpu` `normalWgpuMeshMaterialRenderer.ts`. A
//! lighting-independent debug/utility pass material: `bind` selects the debug
//! pipeline in normal mode, writes the shared Frame uniform (lights ignored), and
//! binds the material's `normal_scale`; `draw` issues the indexed draw. The fragment
//! stage transforms the geometric normal by the normal matrix (WORLD-space) and
//! encodes it as `n * 0.5 + 0.5`. The tangent-space normal map is deferred on wgpu
//! (`has_normal_map` stays false; the placeholder is bound).

use flighthq_materials::normal_material_kind;
use flighthq_render_wgpu::WgpuRenderState;
use flighthq_types::camera::Camera;
use flighthq_types::material::Material;
use flighthq_types::scene_render::{SceneLightBlock, SceneRenderProxy};
use flighthq_types::unlit_material::NormalMaterial;

use crate::wgpu_debug_prelude::{
    WgpuDebugDefineKey, WgpuDebugMode, bind_wgpu_debug_surface, ensure_wgpu_debug_pipeline,
};
use crate::wgpu_mesh_material_registry::{
    WgpuMeshMaterialRenderer, register_wgpu_mesh_material_renderer,
};
use crate::wgpu_mesh_pipeline::{draw_wgpu_mesh_material_subset, write_wgpu_frame_uniform};
use crate::wgpu_scene_runtime::{WgpuMeshUpload, WgpuSceneRuntime};

/// The built-in Normal forward mesh-material renderer.
pub struct NormalWgpuMeshMaterialRenderer;

impl WgpuMeshMaterialRenderer for NormalWgpuMeshMaterialRenderer {
    fn bind(
        &self,
        state: &mut WgpuRenderState,
        scene: &mut WgpuSceneRuntime,
        material: Option<&dyn Material>,
        _lights: &SceneLightBlock,
        camera: &Camera,
    ) {
        let normal = material.and_then(downcast_normal);
        let format = state.runtime.current_color_format.unwrap_or(state.format);
        // The normal map is not yet uploaded on wgpu, so the map variant is never
        // selected — keep has_normal_map false and bind the shared placeholder.
        let key = WgpuDebugDefineKey {
            has_normal_map: false,
            mode: WgpuDebugMode::Normal,
        };
        let cache_key = ensure_wgpu_debug_pipeline(state, scene, &key, format);
        scene.active_pipeline_key = Some(cache_key.clone());
        write_wgpu_frame_uniform(state, scene, camera, _lights);

        let material_key = material.map_or_else(normal_material_kind, |m| m.kind());
        let normal_scale = normal.map_or(1.0, |m| m.normal_scale);
        bind_wgpu_debug_surface(
            state,
            scene,
            &cache_key,
            material_key,
            0.0,
            1.0,
            normal_scale,
        );
        scene.active_material_key = Some(material_key);
    }

    fn draw(
        &self,
        state: &mut WgpuRenderState,
        scene: &mut WgpuSceneRuntime,
        proxy: &SceneRenderProxy,
        upload: &WgpuMeshUpload,
    ) {
        draw_wgpu_mesh_material_subset(state, scene, proxy, upload);
    }
}

/// Returns the singleton Normal renderer instance.
pub fn normal_wgpu_mesh_material_renderer() -> NormalWgpuMeshMaterialRenderer {
    NormalWgpuMeshMaterialRenderer
}

/// Registers the built-in Normal renderer for the Normal material kind on this scene
/// runtime. Opt-in (no top-level side effect).
pub fn register_normal_wgpu_material(state: &mut WgpuRenderState, scene: &mut WgpuSceneRuntime) {
    register_wgpu_mesh_material_renderer(
        state,
        scene,
        normal_material_kind(),
        Box::new(NormalWgpuMeshMaterialRenderer),
    );
}

fn downcast_normal(material: &dyn Material) -> Option<&NormalMaterial> {
    (material as &dyn core::any::Any).downcast_ref::<NormalMaterial>()
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::wgpu_scene_runtime::create_wgpu_scene_runtime;

    mod register_normal_wgpu_material {
        use super::*;

        #[test]
        fn registers_the_renderer_under_the_normal_kind() {
            let mut scene = create_wgpu_scene_runtime();
            scene.material_registry.insert(
                normal_material_kind(),
                Box::new(NormalWgpuMeshMaterialRenderer),
            );
            assert!(
                scene
                    .material_registry
                    .contains_key(&normal_material_kind())
            );
        }
    }
}

File: packages/scene-wgpu/src/normalWgpuMeshMaterialRenderer.ts
import { getWgpuRenderStateRuntime } from '@flighthq/render-wgpu';
import type {
  Camera3D,
  Material,
  MeshGeometry,
  NormalMaterial,
  SceneLightBlock,
  SceneRenderProxy,
  WgpuMeshMaterialRenderer,
  WgpuRenderState,
} from '@flighthq/types';
import { NormalMaterialKind } from '@flighthq/types';

import { bindWgpuDebugSurface, ensureWgpuDebugPipeline } from './wgpuDebugPrelude';
import { registerWgpuMeshMaterialRenderer } from './wgpuMeshMaterialRegistry';
import { beginWgpuMeshDraw, drawWgpuMeshSubset, writeWgpuFrameUniform } from './wgpuMeshPipeline';

// The built-in Normal forward renderer (WgpuMeshMaterialRenderer for NormalMaterialKind) — the WGSL
// mirror of normalGlMeshMaterialRenderer. A lighting-independent debug/utility pass material: bind
// selects the debug pipeline in normal mode for the color format, writes the shared Frame uniform
// (lights ignored), and binds the material's normalScale; draw issues the indexed draw. The fragment
// stage transforms the geometric normal by the normal matrix (so the visualized normal is WORLD-space)
// and encodes it as `n * 0.5 + 0.5` LINEAR color. The tangent-space normal map is NOT sampled on wgpu
// yet — hasNormalMap stays false and the shared placeholder texture is bound (mirrors the documented map
// gap on the rest of the wgpu side; see wgpuDebugPrelude). See registerNormalWgpuMaterial to install it.
export const normalWgpuMeshMaterialRenderer: WgpuMeshMaterialRenderer = {
  bind(
    state: WgpuRenderState,
    material: Readonly<Material> | null,
    _lights: Readonly<SceneLightBlock>,
    camera: Readonly<Camera3D>,
  ): void {
    const stateRuntime = getWgpuRenderStateRuntime(state);
    const pass = stateRuntime.renderPass;
    if (pass === null) return;

    const normal = material as Readonly<NormalMaterial> | null;
    const format = stateRuntime.currentColorFormat ?? state.format;
    // normalMap is not yet uploaded on wgpu, so the map variant is never selected — keep hasNormalMap
    // false and bind the shared placeholder. This will flip to follow normal.normalMap when wgpu texture
    // upload lands, matching the GL renderer's behavior.
    const pipeline = ensureWgpuDebugPipeline(state, { hasNormalMap: false, mode: 'normal' }, format);
    writeWgpuFrameUniform(state, camera, _lights);

    let group: GPUBindGroup;
    if (normal === null) {
      group = bindWgpuDebugSurface(state, pipeline, FALLBACK_MATERIAL, 0, 1, 1);
    } else {
      group = bindWgpuDebugSurface(state, pipeline, normal, 0, 1, normal.normalScale);
    }

    beginWgpuMeshDraw(state, pipeline);
    pass.setBindGroup(2, group);
  },

  draw(state: WgpuRenderState, proxy: Readonly<SceneRenderProxy>, geometry: Readonly<MeshGeometry>): void {
    drawWgpuMeshSubset(state, proxy, geometry);
  },
};

// Registers the built-in Normal renderer for NormalMaterialKind on this state. Opt-in (no top-level side
// effect); call once per WgpuRenderState before drawWgpuScene so meshes with NormalMaterials draw.
export function registerNormalWgpuMaterial(state: WgpuRenderState): void {
  registerWgpuMeshMaterialRenderer(state, NormalMaterialKind, normalWgpuMeshMaterialRenderer);
}

const FALLBACK_MATERIAL = {} as Readonly<NormalMaterial>;
